infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProjectTest.java

```diff
@@ -124,6 +141,69 @@ public void testOpenShiftProjectCleaningUpIfExceptionsOccurs() throws Exception
     verify(routes).delete();
   }
 
+  @Test(expectedExceptions = InfrastructureException.class)
+  public void testThrowsInfrastructureExceptionWhenFailedToGetNamespaceServiceAccounts()
+      throws Exception {
+    prepareProjectRequest();
+    final Resource resource = prepareProjectResource(PROJECT_NAME);
+    doThrow(new KubernetesClientException("error", 403, null)).when(resource).get();
+    doThrow(KubernetesClientException.class).when(openShiftClient).serviceAccounts();
+
+    new OpenShiftProject(clientFactory, PROJECT_NAME, WORKSPACE_ID);
+  }
+
+  @Test(expectedExceptions = InfrastructureException.class)
+  public void testThrowsInfrastructureExceptionWhenServiceAccountEventNotPublished()
+      throws Exception {
+    prepareProjectRequest();
+    final Resource resource = prepareProjectResource(PROJECT_NAME);
+    doThrow(new KubernetesClientException("error", 403, null)).when(resource).get();
+    when(serviceAccountResource.get()).thenReturn(null);
+
+    new OpenShiftProject(clientFactory, PROJECT_NAME, WORKSPACE_ID);
+  }
+
+  @Test(expectedExceptions = InfrastructureException.class)
+  public void testThrowsInfrastructureExceptionWhenWatcherClosed() throws Exception {
+    prepareProjectRequest();
+    final Resource resource = prepareProjectResource(PROJECT_NAME);
+    doThrow(new KubernetesClientException("error", 403, null)).when(resource).get();
+    when(serviceAccountResource.get()).thenReturn(null);
+    doAnswer(
+            (Answer<Watch>)
+                invocation -> {
+                  final Watcher<ServiceAccount> watcher = invocation.getArgument(0);
+                  watcher.onClose(mock(KubernetesClientException.class));
+                  return mock(Watch.class);
+                })
+        .when(serviceAccountResource)
+        .watch(any());
+
+    new OpenShiftProject(clientFactory, PROJECT_NAME, WORKSPACE_ID);
+  }
+
+  @Test
+  public void testStopsWaitingServiceAccountEventJustAfterEventReceived() throws Exception {
```

Comment:
I wouldn't test it twice, here and in KubernetesNamespaceTest. Instead of it, I'd create spy and check that `waitDefaultServiceAccount` method is invoked when a project is created.
WDYT?

Reply:
agree, but since the method `waitDefaultServiceAccount` is protected I'll just mock all needed components and remove duplicated tests.